Cards/CardBase.gd

```diff
@@ -0,0 +1,84 @@
+extends Resource
+class_name CardBase
+## Resource that holds data about a card and provides functions that all derived cards could
+## use if they want.
+##
+## The intention with this is to provide functionality that is common to a lot of cards,
+## like dealing damage, drawing cards, and restoring resources.
+## If you want functionality that's unique to a certain card, then create a new child of this
+## and override one of the functions below.
+## This resource also has data that is used for displaying the card in the world, like
+## description, title, and key art.
+## This includes functionality for applying damage to both the target and caster. Casters may
+## wish to take damage in some contexts.
+## For example, consider the card: "Deal 10 damage to all enemies, but take 3 damage"
+
+
+@export var damage_to_apply_to_target: float = 0.0
+@export var damage_to_apply_to_caster: float = 0.0
+@export var buffs_to_apply_to_target: Array[BuffBase]
+@export var buffs_to_apply_to_caster: Array[BuffBase]
+@export var affect_all_targets: bool = false
+@export var affect_all_casters: bool = false
+@export var application_type: Enums.ApplicationType = Enums.ApplicationType.ENEMY_ONLY
+@export var card_title: String = "NULL"
+@export var card_key_art: ImageTexture = null
+@export var card_description: String = "NULL"
+
+
+func try_play_card(caster: Entity, target: Entity) -> bool:
+	if caster.get_party_component().can_play_on_entity(application_type, target):
+		_on_card_play(caster, target)
+		return true
+	return false
+
+
+func _on_card_play(caster: Entity, target: Entity) -> void:
+	_deal_damage(caster, target)
+	_apply_buffs(caster, target)
+	# TODO add other functionality that lots of cards may share (eg: restore AP, draw cards)
+
+
+# override in child cards if you want to deal damage in a unique way
+func _deal_damage(caster: Entity, target: Entity) -> void:
+	# damage target
+	if damage_to_apply_to_target != 0.0:
+		_damage_entity(target, caster, damage_to_apply_to_target, affect_all_targets)
+	
+	#damage caster
+	if damage_to_apply_to_caster != 0.0:
+		_damage_entity(caster, caster, damage_to_apply_to_caster, affect_all_casters)
+
+
+func _damage_entity(target: Entity, caster: Entity, damage_amount: float, damage_all: bool) -> void:
```

Comment:
it seems to me that `_damage_entity` is the only function (and the ctrl f seems to agree with me) where the signature has the `target` before the `caster`. I think it would be good to swap the position of `target` and `caster` for uniformity's sake

Reply:
Good catch. I guess a good rule of thumb is to always put `caster` before `target` in function params